src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Components.RenderTree
+{
+    /// <summary>
+    /// Represents a range of elements within an instance of <see cref="ArrayBuilder{T}"/>.
+    /// </summary>
+    /// <typeparam name="T">The type of the elements in the array</typeparam>
+    public readonly struct ArrayBuilderSegment<T> : IEnumerable<T>
+    {
+        private readonly ArrayBuilder<T> _builder;
+        private readonly int _offset;
+        private readonly int _count;
+
+        internal ArrayBuilderSegment(ArrayBuilder<T> builder, int offset, int count)
+        {
+            _builder = builder;
+            _offset = offset;
+            _count = count;
+        }
+
+        /// <summary>
+        /// Gets the current underlying array holding the segment's elements.
+        /// </summary>
+        public T[] Array => _builder?.Buffer;
+
+        /// <summary>
+        /// Gets the offset into the underlying array holding the segment's elements.
+        /// </summary>
+        public int Offset => _offset;
+
+        /// <summary>
+        /// Gets the number of items in the segment.
+        /// </summary>
+        public int Count => _count;
+
+        /// <summary>
+        /// Gets the specified item from the segment.
+        /// </summary>
+        /// <param name="index">The index into the segment.</param>
+        /// <returns>The array entry at the specified index within the segment.</returns>
+        public T this[int index]
+            => _builder.Buffer[_offset + index];
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
```

Comment:
If it is common to iterate over this type, we should have a `GetEnumerator` method that returns `ArraySegment.Enumerator`: 

```C#
public ArraySegment<T>.Enumerator GetEnumerator() => new ArraySegment<T>(_builder.Buffer, _offset, _count)).GetEnumerator();
```

The compiler would use duck typing and avoid boxing the struct

Reply:
It's a good idea, but `ArraySegment<T>.Enumerator` is only in `netstandard2.1`, not `netstandard2.0`. I've added a comment here to remind us to consider this if we upgrade in the future.